### PR TITLE
Introduce a recorder demo app built with ScriptProcessorNode

### PR DESCRIPTION
**Pull request by eldrickb**
This PR is the first of a few to introduce snippets for the incoming Audio Worklet Shim Library, a  project to help ease the transition from ScriptProcessorNode to AudioWorkletNode. 

This PR specifically includes a demo of a recording app using ScriptProcessorNode to capture, process, and visualize audio. 

Features 
- Records audio using ScriptProcessorNode
- Presents the recording for playback or download 
- Renders a visualizer for live  and recorded audio 

Questions, known bugs 
- I'm not sure if the AnalyserNode frequency visualizer is necessary 
- Record buffer is set to 5 seconds as a length param is required. 
	- Is there a way to make buffers of arbitrary length?
- Calling prepareClip() after recording is paused causes the app to freeze for a moment
- Sometimes frequency visualizer only shows lowest 16ish frequencies
	- Haven't tested with alternate mics to ensure it's not an isolated issue 
- Some audio seems to artifact / produce weird clicks in the recording 

Preview:
image

### CODE REVIEW

File: src/audio-worklet/shim/spn-demo/app.mjs [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+  visToggle.addEventListener('click', (e) => {
+    visualizationEnabled = !visualizationEnabled;
+    visToggle.querySelector('span').innerHTML = visualizationEnabled ?
+            'Pause' :
+            'Play';
+  });
+}
+
+/**
+ * Script Processor Frequency Visualizer
```
[the remark this answers is not in the document]
Comment (hoch) [reply src/audio-worklet/shim/spn-demo/app.mjs:190]: Also why do you need AnalyserNode when you're doing the visualization with ScriptProcessorNode?

File: src/audio-worklet/shim/spn-demo/app.mjs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+  const bufferLength = analyserNode.frequencyBinCount;
+  const dataArray = new Uint8Array(bufferLength);
+
+  const width = canvas.width;
+  const height = canvas.height;
+  const barWidth = width / bufferLength;
+
+  function draw() {
+    requestAnimationFrame(draw);
```
Comment (hoch) [inline src/audio-worklet/shim/spn-demo/app.mjs:239]: By this way, you're putting multiple render functions into rAF. I think it's better to have 1 main render function and render everything there in the order you intended.

File: src/audio-worklet/shim/spn-demo/exporter.mjs [hunk 3]
```diff
@@ -0,0 +1,137 @@
+/**
```
Comment (hoch) [inline src/audio-worklet/shim/spn-demo/exporter.mjs:1]: Is this the exact copy of the utility function? Or do we have any changes to be reviewed?

**Review by hoch — CHANGES_REQUESTED** [review]
Generally the PR looks great!

Some high-level comments:
- Make comments more meaningful and purposeful. Try to explain why rather than what.
- Each file needs an overview comment.
- Each function needs an overview comment.
- The Chromium convention is not to shorten variable names. Make them as descriptive as possible while keeping it short. Find a balance between extremes.

Let me know if you have any questions!

File: src/audio-worklet/shim/spn-demo/exporter.mjs (hunk 3 above)
Comment (eldrickb) [reply src/audio-worklet/shim/spn-demo/exporter.mjs:1]: Exact copy, pulled outside of the wrapping class

File: src/audio-worklet/shim/spn-demo/app.mjs [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+      noiseSuppression: false,
+      latency: 0,
+    },
+  });
+  const micSourceNode = context.createMediaStreamSource(micStream);
+
+  // Setup Script Processor Node
+  const recordBuffer = context.createBuffer(
+      2,
+      context.sampleRate * 5,
```
[the remark this answers is not in the document]
Comment (eldrickb) [reply src/audio-worklet/shim/spn-demo/app.mjs:37]: Forgot to ask, Although the length param is required, is there a way to make a limitless buffer?

File: src/audio-worklet/shim/spn-demo/app.mjs [hunk 5]
```diff
@@ -0,0 +1,307 @@
+'use strict';
```
[the remark this answers is not in the document]
Comment (eldrickb) [reply src/audio-worklet/shim/spn-demo/app.mjs:1]: .mjs means Module JS, the paradigm for importing/exporting modules as we know it sing "import"/"export default" terms. But this obviously isn't a module anymore and just a general script, so I'll change it back to .js

File: src/audio-worklet/shim/spn-demo/app.mjs (hunk 1 above)
[the remark this answers is not in the document]
Comment (eldrickb) [reply src/audio-worklet/shim/spn-demo/app.mjs:190]: Something I had in the demo from the beginning and never removed. I was questioning the value of it being included myself. I myself think it's not really relevant to demonstrating SPN anymore and should probably just remove it. Your thoughts?

File: src/audio-worklet/shim/spn-demo/exporter.mjs (hunk 3 above)
Comment (hoch) [reply src/audio-worklet/shim/spn-demo/exporter.mjs:1]: ACK

File: src/audio-worklet/shim/spn-demo/index.njk [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  title: ScriptProcessor Recording App
+  parent: audio-worklet
+to_root_dir: ../../../
+---
+
+{% extends "../../../_includes/base.njk" %}
+
+{% block content %}
+
+<style>
```
Comment (hoch) [inline src/audio-worklet/shim/spn-demo/index.njk:13]: style: 2-space indentation

**Review by hoch — COMMENTED** [review]
Overall looks great! There still are some comments to be addressed, but I think we're close.

**Comment by eldrickb** [comment]
Ready for next review!

File: src/audio-worklet/migration/spn-recorder/README.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# ScriptProcessorNode Demo
+
+This demonstrates a basic web audio recording app using ScriptProcessorNode.
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/README.md:3]: Remove "web audio".

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 8]
```diff
@@ -0,0 +1,289 @@
+/**
+ * @license Copyright (c) Google. MIT License.
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:2]: We're not using MIT license:
https://github.com/GoogleChromeLabs/web-audio-samples/blob/main/LICENSE

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+
+// Arbitrary buffer size, not specific for a reason
+const BUFFER_SIZE = 256;
+let isRecording = false;
+let isMonitoring = false;
+let visualizationEnabled = true;
+let currentSample = [];
+let recordingLength = 0;
+
+// Wait for user interaction to initialize audio, as per specification
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:20]: nit: A missing period at the end.

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+document.addEventListener('click', (element) => {
+  init();
+  document.querySelector('#click-to-start').remove();
+}, {once: true});
+
+/**
+ * Defines overall audio chain and initializes all functionality
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:27]: nit: missing period.

Comment (hoch) [reply src/audio-worklet/migration/spn-recorder/app.js:27]: I also saw a lot of places without a proper period. Let's fix them all.

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+/**
+ * Creates ScriptProcessor to record and track microphone audio.
+ * @param {AudioBuffer} recordBuffer
+ * @return {ScriptProcessorNode} ScriptProcessorNode to pass audio into.
+ */
+function setupScriptProcessor(recordBuffer) {
+  const processor = context.createScriptProcessor(BUFFER_SIZE);
+  currentSample = new Array(recordBuffer.numberOfChannels).fill([]);
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:78]: Didn't notice that this is a global - this example is supposed to be "bad" so it shouldn't matter much, but how can we change this so it doesn't live on a global scope?

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+
+  // Main SPN callback. Handles recording data and tracking recording length
+  processor.onaudioprocess = function(event) {
+    // Perform logic on all mic channels, regardless of setup
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:82]: All "input" channel - the incoming signal could be anything.

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+import createLinkFromAudioBuffer from './exporter.mjs';
+
+const context = new AudioContext();
+
+// Arbitrary buffer size, not specific for a reason
+const BUFFER_SIZE = 256;
+let isRecording = false;
+let isMonitoring = false;
+let visualizationEnabled = true;
+let currentSample = [];
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:17]: Just to clarify that it is an array with multiple values: `currentSamples`

File: src/audio-worklet/migration/spn-recorder/index.njk [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+  title: ScriptProcessor Recording App
+  parent: audio-worklet
+to_root_dir: ../../../
+---
+
+{% extends "../../../_includes/base.njk" %}
+
+{% block content %}
+
+<style>
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/index.njk:13]: Use 2-space indentation.

File: src/audio-worklet/migration/spn-recorder/index.njk [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+					<button>download file</button>
+				</a>
+			</div>
+		</div>
+	</div>
+</div>
+
+<script type="module" src="app.js"></script>
+
+{% endblock %}
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/index.njk:71]: Add a new line character at the EOF.

**Review by hoch — COMMENTED** [review]
It's getting there, but I still think we need to fix some comments.

- If the comment is not explaining "why", please reconsider its meaning.
- Let's fix up all grammar and punctuation!
- Check the coding style: 80-col line wrapping and indentation.

File: src/audio-worklet/migration/spn-recorder/app.js (hunk 8 above)
Comment (eldrickb) [reply src/audio-worklet/migration/spn-recorder/app.js:2]: Just gonna use the same header as the rest of the files

File: src/audio-worklet/migration/spn-recorder/app.js (hunk 11 above)
Comment (eldrickb) [reply src/audio-worklet/migration/spn-recorder/app.js:78]: Replaced the global with a passthrough function

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 16]
```diff
@@ -0,0 +1,306 @@
+// Copyright (c) 2021 The Chromium Authors. All rights reserved.
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:1]: s/2021/2022

File: src/audio-worklet/migration/spn-recorder/README.md [hunk 17]
```diff
@@ -0,0 +1,3 @@
+# ScriptProcessorNode Demo
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/README.md:1]: How about "Recorder demo using ScriptProcessorNode"?

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+    player.src = wavUrl;
+    downloadButton.src = wavUrl;
+    downloadButton.download = 'recording.wav';
+
+    // Display current recording length.
+    document.querySelector('#data-len').innerHTML =
+        Math.round(recordingLength / recordBuffer.sampleRate * 100)/100;
+  }
+
+  recordButton.addEventListener('click', (e) => {
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:135]: nit: `e` to `event`.

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+    currentSamples = newSamples;
+
+    if (!firstSamplesReceived) {
+      draw();
+      firstSamplesReceived = true;
+    }
+  };
+
+  const visToggle = document.querySelector('#viz-toggle');
+  visToggle.addEventListener('click', (e) => {
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:189]: nit: `e` to `event`.

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 20]
```diff
@@ -0,0 +1,306 @@
...
+    visualizationEnabled = !visualizationEnabled;
+    visToggle.querySelector('span').innerHTML =
+      visualizationEnabled ? 'Pause' : 'Play';
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:192]: In this case, the indentation should be 4 spaces.

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 21]
```diff
@@ -0,0 +1,306 @@
...
+      // We only need to calculate this value once per render frame.
+      let currentSampleGain = 0;
+
+      for (let i = 0; i < currentSamples.length; i++) {
+        for (let j = 0; j < currentSamples[i].length; j++) {
+          currentSampleGain += currentSamples[i][j];
+        }
+      }
+
+      currentSampleGain /= (currentSamples.length *currentSamples[0].length);
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:207]: nit: a space after `*`

File: src/audio-worklet/migration/spn-recorder/app.js [hunk 22]
```diff
@@ -0,0 +1,306 @@
...
+  const canvas = document.querySelector('#recording-canvas');
+  const canvasContext = canvas.getContext('2d');
+
+  const width = canvas.width;
+  const height = canvas.height;
+
+  canvasContext.fillStyle = 'red';
+  canvasContext.fillRect(0, 0, 1, 1);
+
+  let currX = 0;
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/app.js:273]: nit: either `currentX` or `x`.

File: src/audio-worklet/migration/spn-recorder/index.njk [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+  title: ScriptProcessor Recording App
+  parent: audio-worklet
+to_root_dir: ../../../
+---
+
+{% extends "../../../_includes/base.njk" %}
+
+{% block content %}
+
+  <style>
```
Comment (hoch) [inline src/audio-worklet/migration/spn-recorder/index.njk:13]: We can remove the 2-space indentation in the entire scope.

**Review by hoch — APPROVED** [review]
LGTM, but please fix nits before merging.

**Comment by eldrickb** [comment]
fixed! though @hoch you'll need to hit the button, I'm not authorized

**Pull request merged**